### PR TITLE
reverseproxy: Implement `weighted_round_robin` load balancing policy

### PR DESCRIPTION
**Pull request by sabify**

### CODE REVIEW

File: modules/caddyhttp/reverseproxy/selectionpolicies_test.go [hunk 1]
```diff
@@ -74,6 +74,63 @@ func TestRoundRobinPolicy(t *testing.T) {
...
+	}
+	req, _ := http.NewRequest("GET", "/", nil)
+
+	h := wrrPolicy.Select(pool, req, nil)
+	if h != pool[0] {
+		t.Error("Expected first weighted round robin host to be first host in the pool.")
+	}
+	h = wrrPolicy.Select(pool, req, nil)
+	if h != pool[0] {
+		t.Error("Expected second weighted round robin host to be firest host in the pool.")
```
Comment (francislavoie) [inline modules/caddyhttp/reverseproxy/selectionpolicies_test.go:91]:
```suggestion
		t.Error("Expected second weighted round robin host to be first host in the pool.")
```

**Comment by francislavoie** [comment]
Please add an adapt test to https://github.com/caddyserver/caddy/tree/master/caddytest/integration/caddyfile_adapt to show the different ways to configure it with the Caddyfile

**Comment by sabify** [comment]
Done.

File: caddytest/integration/caddyfile_adapt/reverse_proxy_load_balance_wrr.txt [hunk 2]
```diff
@@ -0,0 +1,71 @@
+:8884
+
+reverse_proxy 127.0.0.1:65535 127.0.0.1:35535 {
+	lb_policy weighted_round_robin "10, 1"
```
Comment (francislavoie) [inline caddytest/integration/caddyfile_adapt/reverse_proxy_load_balance_wrr.txt:4]: I find this syntax kinda strange and not idiomatic - typically arguments in the Caddyfile are space separated, not comma separated. You can use `d.RemainingArgs()` to get a slice of tokens from the same-line.

Comment (sabify) [reply caddytest/integration/caddyfile_adapt/reverse_proxy_load_balance_wrr.txt]: Done.

File: modules/caddyhttp/reverseproxy/selectionpolicies.go [hunk 3]
```diff
@@ -78,6 +79,108 @@ func (r *RandomSelection) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
 	return nil
 }
 
+// WeightedRoundRobinSelection is a policy that selects
+// a host based on weighted round-robin ordering.
+type WeightedRoundRobinSelection struct {
+	// The weight of each upstream in order.
```
Comment (mholt) [inline modules/caddyhttp/reverseproxy/selectionpolicies.go:85]:
```suggestion
	// The weight of each upstream in order, corresponding with the list of upstreams configured.
```

File: modules/caddyhttp/reverseproxy/selectionpolicies.go [hunk 4]
```diff
@@ -78,6 +79,108 @@ func (r *RandomSelection) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
...
+}
+
+// Provision sets up r.
+func (r *WeightedRoundRobinSelection) Provision(ctx caddy.Context) error {
+	if len(r.Weights) != 0 {
+		r.totalWeight = 0
+		for _, weight := range r.Weights {
+			r.totalWeight += weight
+		}
+	}
```
Comment (mholt) [inline modules/caddyhttp/reverseproxy/selectionpolicies.go:131]:
```suggestion
	for _, weight := range r.Weights {
		r.totalWeight += weight
	}
```

0 is the default value anyway, and no iteration will happen if len is 0. Simpler code :)

File: modules/caddyhttp/reverseproxy/selectionpolicies.go [hunk 5]
```diff
@@ -78,6 +79,108 @@ func (r *RandomSelection) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
...
+
+		for _, weight := range args {
+			weightInt, err := strconv.Atoi(weight)
+			if err != nil {
+				return d.Errf("invalid weight value '%s': %v", weight, err)
+			}
+			if weightInt < 1 {
+				return d.Errf("invalid weight value '%s': weight should be non-zero and positive", weight)
+			}
+			r.totalWeight += weightInt
```
Comment (mholt) [inline modules/caddyhttp/reverseproxy/selectionpolicies.go:117]:
```suggestion
```

This is done in Provision

File: modules/caddyhttp/reverseproxy/selectionpolicies.go [hunk 6]
```diff
@@ -78,6 +79,108 @@ func (r *RandomSelection) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
...
+		totalWeight := 0
+		for _, weight := range r.Weights {
+			totalWeight += weight
+		}
+		if r.totalWeight != totalWeight {
+			return fmt.Errorf("total weight miscalculated")
+		}
+	}
+	return nil
+}
```
Comment (mholt) [inline modules/caddyhttp/reverseproxy/selectionpolicies.go:147]: Not sure what this is for :thinking:

File: modules/caddyhttp/reverseproxy/selectionpolicies.go [hunk 7]
```diff
@@ -78,6 +79,108 @@ func (r *RandomSelection) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
...
+// Select returns an available host, if any.
+func (r *WeightedRoundRobinSelection) Select(pool UpstreamPool, _ *http.Request, _ http.ResponseWriter) *Upstream {
+	if len(pool) == 0 {
+		return nil
+	}
+	if len(r.Weights) < 2 {
+		return pool[0]
+	}
+	index := 0
+	totalWeight := 0
```
Comment (mholt) [inline modules/caddyhttp/reverseproxy/selectionpolicies.go:158]:
```suggestion
	var index, totalWeight int
```

Totally just a style nit, but hear me out :)

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for the contribution!

I have just a few nits that I think could be addressed, and then if updated today I can get this in for beta 2.

**Review by mholt — APPROVED** [review]
Looks good now!! Thanks so much!

**Pull request merged**